proposals/text/6001-document-store-filter-rework.md

```diff
@@ -0,0 +1,197 @@
+- Title: Document Stores filter specification for Haystack 2.x
+- Decision driver: Silvano Cerza
+- Start Date: 2023-10-08
+- Proposal PR: [#6001](https://github.com/deepset-ai/haystack/pull/6001)
+
+# Summary
+
+This proposal introduces a new fully detailed and extensible specification for filtering Document Stores in Haystack 2.x.
+This comes from my personal experience and struggles trying to implement filters conversion for `ElasticsearchDocumentStore`.
+
+# Basic example
+
+```
+{
+  "conditions": [
+    { "field": "age", "operator": ">=", "value": 18 },
+    {
+      "operator": "OR",
+      "conditions": [
+        { "field": "country", "operator": "==", "value": "USA" },
+        { "field": "country", "operator": "==", "value": "Canada" }
+      ]
+    }
+  ]
+}
+```
+
+The above snippet would be equivalent to the following Python code:
+
+```python
+age >= 18 and (country == "USA" or country == "Canada)
+```
+
+# Motivation
+
+Filtering in Haystack 1.x has no detailed clear specification, we only have an [high level overview][filters-high-level-doc] in the documentation that also mixes REST APIs documentation in. It's also inherited from MongoDB and is a subset of the Mongo Query Language.
+
+Converting from the current filters to another query language is quite hard as there are tons of corner cases. Handling nested filters is usually really error prone as the operators can be keys, this requires ton of nested and/or recursive logic to figure out whether the current key is a field or an operator. There's also quite some backtracking involved as the field could be two or three levels above its comparison operator and/or value.
+
+As a practical example the below two filters are equivalent. Given that they're structured differently and `$and` is implicit in the second one:
+
+```
+{"number": {"$and": [{"$lte": 2}, {"$gte": 0}]}}
+
+{"number": {"$lte": 2, "$gte": 0}}
+```
+
+With the newly proposed approach the first filter would be equivalent to:
+
+```
+{
+    "operator": "AND",
+    "conditions": [
```

Comment:
Do we need this `conditions` key? Can't we put `AND` there and get rid of `operator` as well? This would become:
```
{
    "AND": [
        { "field": "number", "operator": "<=", "value": 2 },
        { "field": "number", "operator": ">=", "value": 0 },
    ]
}
```

Reply:
That's a possibility but I feel this was one of the pitfalls of the previous design. 

I think it makes the logic less straightforward as you need to search for the operator every time to figure out which kind of dictionary it is.

Having operator and conditions separate makes it easier to figure out the filters I think.